### PR TITLE
Feature/synchronizer

### PR DESCRIPTION
**Pull request by christophercampbell**
This PR adds a data synchronizer. 

1. sync_info table for tracking last processed block for SequenceBatches
2. DataCommitteeWatcher for data committee membership changes
3. BatchSynchronizer that watches for SequenceBatches events starting from last block in sync_info
  - checks if keys in batch are stored
  - retrieves values for missing keys from other data committee members
  - stores resolved offchain data & sync_info

### CODE REVIEW

**Review by arnaubennassar — CHANGES_REQUESTED** [review]

File: db/db.go [hunk 1]
```diff
@@ -60,11 +61,55 @@ func (p *DB) GetOffChainData(ctx context.Context, key common.Hash, dbTx pgx.Tx)
...
 
+	println(key.Hex())
+
 	if err := dbTx.QueryRow(ctx, getOffchainDataSQL, key.Hex()).Scan(&valueStr); err != nil {
 		if errors.Is(err, pgx.ErrNoRows) {
 			return nil, state.ErrStateNotSynchronized
 		}
 		return nil, err
 	}
+	valueStr = strings.TrimPrefix(valueStr, "0x") // is this right?
```
[the remark this answers is not in the document]
Comment (christophercampbell) [reply db/db.go:72]: What is the right way to decode the hex? `common.HexToBytes` fails if the '0x' is included. maybe `common.FromHex`?

File: synchronizer/batches.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+	return nil
+}
+
+func rollback(ctx context.Context, err error, dbTx pgx.Tx) {
+	if txErr := dbTx.Rollback(ctx); txErr != nil {
+		log.Errorf("failed to roll back transaction after error %v : %v", err, txErr)
+	}
+}
+
+func (bs *BatchSynchronizer) resolve(key common.Hash) (offchaindata.OffChainData, error) {
```
[the remark this answers is not in the document]
Comment (christophercampbell) [reply synchronizer/batches.go:185]: No random is the committee list permuted, the len(random) is the length of the list, not a random index into list. So it should be named better, maybe `shuffled`.

It stops once it finds the data, and will potentially try each. 

Skipping self is good idea.

File: db/db.go [hunk 3]
```diff
@@ -60,11 +61,55 @@ func (p *DB) GetOffChainData(ctx context.Context, key common.Hash, dbTx pgx.Tx)
...
+
+// Exists checks if a key exists in offchain data table
+func (db *DB) Exists(ctx context.Context, key common.Hash) bool {
+	var keyExists = "SELECT COUNT(*) FROM data_node.offchain_data WHERE key = $1"
+	var (
+		count uint
+	)
+
+	if err := db.pg.QueryRow(ctx, keyExists, key.Hex()).Scan(&count); err != nil {
+		return false
```
[the remark this answers is not in the document]
Comment (christophercampbell) [reply db/db.go:84]: Retrying on DB errors can be tricky, there are classes of errors that won't resolve with retry, so can get into a loop. I guess maybe have a fixed number of retries, but then what is the behavior if you reach that?

File: synchronizer/batches.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+func (bs *BatchSynchronizer) getStartBlock() (uint64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), dbTimeout)
+	defer cancel()
+
+	start, err := bs.db.GetLastProcessedBlock(ctx)
+	if err != nil {
+		log.Errorf("error retrieving last processed block, starting from 0: %v", err)
+	}
+	if start > 0 {
+		start = start - 1 // since a block may have been partially processed
```
[the remark this answers is not in the document]
Comment (christophercampbell) [reply synchronizer/batches.go:111]: Can there be more than one event per block? The synchronizer is processing events, so might not have processed all the events in the block.

File: synchronizer/batches.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+		for err != nil {
+			<-time.After(bs.retry)
+			sub, err = bs.client.Supernets2.WatchSequenceBatches(opts, events, nil)
+			if err != nil {
+				log.Errorf("error subscribing to sequence batch events, retrying: %v", err)
+			}
+		}
+
+		// wait on events, timeouts, and signals to stop
+		select {
```
[the remark this answers is not in the document]
Comment (christophercampbell) [reply synchronizer/batches.go:80]: it is, see: 

https://github.com/0xPolygon/supernets2-data-availability/blob/412f5641eb17de0f8da1dfc93cf72414bfd7052f/synchronizer/batches.go#L53

File: synchronizer/committee.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/0xPolygon/supernets2-data-availability/config"
+	"github.com/0xPolygon/supernets2-node/etherman"
+	"github.com/0xPolygon/supernets2-node/etherman/smartcontracts/supernets2datacommittee"
+	"github.com/0xPolygon/supernets2-node/log"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+// DataCommitteeTracker tracks changes to the data committee membership
+type DataCommitteeTracker struct {
```
[the remark this answers is not in the document]
Comment (christophercampbell) [reply synchronizer/committee.go:17]: Good suggestion, much better

**Pull request merged**